### ISSUE TITLE
[Tracking issue] Sort sometime hangs when running on large fields

### ISSUE THREAD

**Issue by akoshelev** [private-attribution/ipa#237]
This is the tracking issue for sporadic failures I am observing while running the following program (bench). This is almost an exact copy of `generate_sort_permutation` test, but rans faster under bench profile and therefore have a higher chance to be reproduced:

```rust
use std::time::Instant;
use futures_util::future::try_join_all;
use rand::Rng;
use raw_ipa::error::BoxError;
use raw_ipa::ff::Fp32BitPrime;
use raw_ipa::protocol::QueryId;
use raw_ipa::test_fixture::{make_contexts, make_world_with_config, TestWorldConfig, validate_and_reconstruct};
use raw_ipa::protocol::sort::generate_sort_permutation::GenerateSortPermutation;
use raw_ipa::ff::Field;

#[tokio::main(flavor = "multi_thread", worker_threads = 3)]
async fn main() -> Result<(), BoxError> {
    let mut config = TestWorldConfig::default();
    config.gateway_config.send_buffer_config.items_in_batch = 1;
    config.gateway_config.send_buffer_config.batch_count = 1000;
    let world = make_world_with_config(QueryId, config);
    let [ctx0, ctx1, ctx2] = make_contexts::<Fp32BitPrime>(&world);
    let num_bits = 64;
    let mut rng = rand::thread_rng();

    let batchsize = 100;

    let mut match_keys: Vec<u64> = Vec::new();
    for _ in 0..batchsize {
        match_keys.push(rng.gen::<u64>());
    }

    let input_len = match_keys.len();
    let mut shares = [
        Vec::with_capacity(input_len),
        Vec::with_capacity(input_len),
        Vec::with_capacity(input_len),
    ];
    for match_key in match_keys.clone() {
        let share_0 = rng.gen::<u64>();
        let share_1 = rng.gen::<u64>();
        let share_2 = match_key ^ share_0 ^ share_1;

        shares[0].push((share_0, share_1));
        shares[1].push((share_1, share_2));
        shares[2].push((share_2, share_0));
    }

    let start = Instant::now();
    let result = try_join_all(vec![
        GenerateSortPermutation::new(&shares[0], num_bits).execute(ctx0),
        GenerateSortPermutation::new(&shares[1], num_bits).execute(ctx1),
        GenerateSortPermutation::new(&shares[2], num_bits).execute(ctx2),
    ]).await?;
    let duration = start.elapsed().as_secs_f32();
    println!("benchmark complete after {duration}s");

    assert_eq!(result[0].len(), input_len);
    assert_eq!(result[1].len(), input_len);
    assert_eq!(result[2].len(), input_len);

    let mut mpc_sorted_list: Vec<u128> = (0..input_len).map(|i| i as u128).collect();
    for (i, match_key) in match_keys.iter().enumerate() {
        let index = validate_and_reconstruct((result[0][i], result[1][i], result[2][i]));
        mpc_sorted_list[index.as_u128() as usize] = u128::from(*match_key);
    }

    let mut sorted_match_keys = match_keys.clone();
    sorted_match_keys.sort_unstable();
    for i in 0..input_len {
        assert_eq!(u128::from(sorted_match_keys[i]), mpc_sorted_list[i]);
    }

    Ok(())
}
```

Noticeable difference is `batch_count` setting which is set to 1000. That allows larger buffers but still given that `batch_size` is set to 1, Infra sends messages one-by-one. There should be no choking in the system with these settings, but it is clearly happens at some point

**Comment by akoshelev** [comment]
Last meaningful messages from `H3` are:

```
2022-11-16T06:29:57.079540Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new ReceiveRequest(channel[peer=H2,step=step=protocol/sort35/mod_conv/xor2], RecordId(96))
2022-11-16T06:29:57.079556Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new ReceiveRequest(channel[peer=H2,step=step=protocol/sort35/mod_conv/xor2], RecordId(97))
2022-11-16T06:29:57.079597Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 72 message(s) from channel[peer=H1,step=step=protocol/sort35/apply_inv/shuffle_inputs/shuffle2]
2022-11-16T06:29:57.079630Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new ReceiveRequest(channel[peer=H2,step=step=protocol/sort35/mod_conv/xor2], RecordId(98))
2022-11-16T06:29:57.079656Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 message(s) from channel[peer=H2,step=step=protocol/sort35/apply_inv/shuffle_inputs/shuffle1]
2022-11-16T06:29:57.079668Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 message(s) from channel[peer=H1,step=step=protocol/sort35/apply_inv/shuffle_inputs/shuffle2]
2022-11-16T06:29:57.079680Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new ReceiveRequest(channel[peer=H2,step=step=protocol/sort35/mod_conv/xor2], RecordId(99))
```

so it is still expecting modulus conversion messages to arrive

**Comment by martinthomson** [comment]
Could you check in the bench?  No sense in going this alone.

**Comment by akoshelev** [comment]
Looking at `step=protocol/sort35/mod_conv/xor2], RecordId(99)` that `H3` is expecting from `H2`. It was sent by `H2` 


```bash
cat -p /tmp/trace.log | grep "protocol/sort35/mod_conv/xor2" | grep "role=H2" | tail -n 2
2022-11-16T06:29:57.057895Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: new SendRequest(channel[peer=H3,step=step=protocol/sort35/mod_conv/xor2], RecordId(99)
2022-11-16T06:29:57.057933Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: sending 8 bytes to channel[peer=H3,step=step=protocol/sort35/mod_conv/xor2]
```


06:29:57.05 message sent from H2 to H3
06:29:57.07 delivery requested by H3

**Comment by akoshelev** [comment]
I am going to add some more traces to help me figure out what is going on here. In particular, I want to see the status of each receive request - whether it was fulfilled immediately or scheduled. Trying to chase down the first unsuccessful receive request that caused the helper to start lagging behind

**Comment by akoshelev** [comment]
only the last record in xor2 step wasn't immediately received by H3

```bash
 cat /tmp/trace2.log| grep role=H3 | grep "step=protocol/sort15/mod_conv/xor2" | grep "ReceiveRequest("
2022-11-16T18:49:35.827846Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new "ReceiveRequest(channel[peer=H2,step=step=protocol/sort15/mod_conv/xor2], RecordId(0))". FulfilledImmediately
2022-11-16T18:49:35.832729Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new "ReceiveRequest(channel[peer=H2,step=step=protocol/sort15/mod_conv/xor2], RecordId(1))". FulfilledImmediately
... (same thing)
"ReceiveRequest(channel[peer=H2,step=step=protocol/sort15/mod_conv/xor2], RecordId(98))". FulfilledImmediately
2022-11-16T18:49:35.859165Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: new "ReceiveRequest(channel[peer=H2,step=step=protocol/sort15/mod_conv/xor2], RecordId(99))". Scheduled

**Comment by akoshelev** [comment]
Ignore the different bits in steps, I am just re-running the bench over and over and it gets stuck in different steps.

I added tracing to track record identifiers received by H3 and there is no evidence that record 99 was received (also ignore "8 messages" it should be read as 8 bytes.

```bash
cat /tmp/trace2.log| grep role=H3 | grep "protocol/sort2/mod_conv/xor2" | grep "peer=H2,step=step=protocol/sort2/mod_conv/xor2" | grep "received"

2022-11-16T19:20:29.915812Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 0..1) message(s) from channel[peer=H2,step=step=protocol/sort2/mod_conv/xor2].
...
2022-11-16T19:20:29.930928Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 98..99) message(s) from channel[peer=H2,step=step=protocol/sort2/mod_conv/xor2].
```

compare this with previous step:

```bash
 cat /tmp/trace2.log| grep role=H3 | grep "protocol/sort1/mod_conv/xor2" | grep "peer=H2,step=step=protocol/sort1/mod_conv/xor2" | grep "received"

2022-11-16T19:20:29.631743Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 0..1) message(s) from channel[peer=H2,step=step=protocol/sort1/mod_conv/xor2].
...
2022-11-16T19:20:29.647567Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 99..100) message(s) from channel[peer=H2,step=step=protocol/sort1/mod_conv/xor2].
```

**Comment by akoshelev** [comment]
One may suspect that control loop is blocked (because of https://github.com/private-attribution/ipa/blob/6d19de7cf3b0f59f1fe6bed3edd8e7be12222279/src/helpers/messaging.rs#L179) but I don't see any evidence of that. After `19:20:29.64` there are plenty of messages logged by `H3` from the control loop

```bash
2022-11-16T19:20:29.952315Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 59..60) message(s) from channel[peer=H2,step=step=protocol/sort2/apply_inv/shuffle_inputs/shuffle1].
2022-11-16T19:20:29.952387Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received 8 (record ids 33..34) message(s) from channel[peer=H1,step=step=protocol/sort2/apply_inv/shuffle_inputs/shuffle2].
```

**Comment by akoshelev** [comment]
For the record, I connected tokio-console but it does not show the stack frame

![image](https://user-images.githubusercontent.com/1110516/202280783-4c9ba557-553e-4930-97e4-bcbec6cb57e6.png)

I would need to look one layer below Infra, which is network. I **don't** believe it is stuck because other messages are being delivered to H3 but better check that

**Comment by akoshelev** [comment]
I don't believe we have a contention on network layer. There is simply no evidence of that. Right now I am working on annotating each `MessageChunks` instance with a unique id (per channel) to see where the message is lost. Two possibilities: either sender is not driving the future to completion or receiver puts it in the wrong place.

**Comment by akoshelev** [comment]
All right, it is just an bug in offset calculation somewhere. Such a relief.

Here is some evidence:

Record 99 is assigned to the chunk 98 by helper H2

```bash
2022-11-16T23:05:48.928336Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: new SendRequest(channel[peer=H3,step=step=protocol/sort18/mod_conv/xor2], RecordId(99)
2022-11-16T23:05:48.928390Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: Sending chunk 98 to channel[peer=H3,step=step=protocol/sort18/mod_conv/xor2]
```

however, receiver thinks it is record 98

```bash
2022-11-16T23:05:48.928502Z TRACE gateway_loop{role=H3}: raw_ipa::helpers::messaging: received chunks [97, 98] and assigned range 97..99 to it, from channel[peer=H2,step=step=protocol/sort18/mod_conv/xor2].
```

**Comment by martinthomson** [comment]
![image](https://user-images.githubusercontent.com/67641/202319169-e1482fad-9c70-4a32-8611-2482afef61f6.png)

**Comment by akoshelev** [comment]
I think I know what is going on and it is a reordering bug in infra. 

This is what triggered this behavior

```bash
2022-11-16T23:05:48.922198Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: new SendRequest(channel[peer=H3,step=step=protocol/sort18/mod_conv/xor2], RecordId(47)
2022-11-16T23:05:48.922225Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: new SendRequest(channel[peer=H3,step=step=protocol/sort18/mod_conv/xor2], RecordId(46)
2022-11-16T23:05:48.922264Z TRACE gateway_loop{role=H2}: raw_ipa::helpers::messaging: Sending chunk 46 to channel[peer=H3,step=step=protocol/sort18/mod_conv/xor2]
```

Records 46 and 47 arrived out of order, so adding record 47 did not trigger send. Adding 46 triggered sending a chunk of **only one record 46** because that is how `SendBuffer` works

https://github.com/private-attribution/ipa/blob/9c4966cbf01bdf97672d37e3b67eb99fced0522a/src/helpers/buffers/fsv.rs#L78

record 46 was sent, but record 47 was left in the buffer. now every send triggers sending previously added message and not the one that was added. After all records have been processed and assuming there was no reordering, one record remains in the buffer because `Gateway` does not drain the send buffer entirely

https://github.com/private-attribution/ipa/blob/6d19de7cf3b0f59f1fe6bed3edd8e7be12222279/src/helpers/messaging.rs#L177

How to fix it? I am inclined to change the contract of 'FixedSizeByteVec::take` method to return everything from the head till the first hole. In other words, if batch_size = 1 and `FixedSizeByteVec` contains 10 elements of 8 byte each, return 80 bytes, instead of 8. That would make chunks non uniformly sized but I think it should be OK. It is certainly more efficient to drain the buffer this way and we **really** want to fix it, we can deal with it in either Infra or Network layer


Obviously I am going to write a unit test that reproduces this issue, seems quite easy to trigger this behavior - just set the batch size to 1 and add elements out of order

**Comment by akoshelev** [comment]
fixed by #252
